### PR TITLE
Bitmap index performance results

### PR DESCRIPTION
**Pull request by lee-lei**
## What changes were proposed in this pull request?

Based on #349 and #350, three metrics are added. All of them are significantly reduced.
1. index file size.
2. query time.
3. index creation time.

## How was this patch tested?

mvn clean test package

### CODE REVIEW

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/BitMapScanner.scala [hunk 1]
```diff
@@ -37,9 +41,9 @@ private[oap] case class BitMapScanner(idxMeta: IndexMeta) extends IndexScanner(i
 
   override def canBeOptimizedByStatistics: Boolean = true
 
-  @transient var internalItr: Iterator[Int] = Iterator[Int]()
+  @transient var internalItr: Iterator[Integer] = Iterator[Integer]()
```
Comment (LuciferYang) [inline src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/BitMapScanner.scala]: Int.scala equivalent to Java's `int` primitive type,not java.lang.Integer，Integer Type will use about 4 times memory than primitive int, and there will more autobox operations.

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/BitMapScanner.scala [hunk 2]
```diff
@@ -114,21 +119,31 @@ private[oap] case class BitMapScanner(idxMeta: IndexMeta) extends IndexScanner(i
...
-        Array.empty[BitSet]
+        Array.empty[ImmutableRoaringBitmap]
       } else {
-        sortedKeyList.slice(startIdx, endIdx + 1).map(key =>
-          hashMap.get(key).get)
+        val partialRbList = new mutable.ListBuffer[ImmutableRoaringBitmap]()
+        val rbStartIdxOffset = Platform.getInt(baseObj, rbOffsetOffset + startIdx * 4)
+        var curPosition = rbStartIdxOffset
+        indexBb.position(curPosition)
+        (startIdx until endIdx + 1).map( idx => {
```
Comment (chenghao-intel) [inline src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/BitMapScanner.scala]:
```scala
scala> 1 until 4
res0: scala.collection.immutable.Range = Range(1, 2, 3)

scala> 1 to 4
res1: scala.collection.immutable.Range.Inclusive = Range(1, 2, 3, 4)
```

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/BitmapIndexRecordWriter.scala [hunk 3]
```diff
@@ -59,28 +65,65 @@ private[index] class BitmapIndexRecordWriter(
...
-    out.writeObject(hashMap)
-    out.flush()
-    val objLen = writeBuf.size()
-    // write byteArray length and byteArray
-    IndexUtils.writeInt(writer, objLen)
-    writer.write(writeBuf.toByteArray)
-    out.close()
-    val indexEnd = 4 + objLen + header
-    val offset: Long = indexEnd
+    // Serialize sortedKeyList and get length
```
Comment (chenghao-intel) [inline src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/BitmapIndexRecordWriter.scala]: Sorry I am not so sure about the index file layout, but I would suggest to add comments for this, and, probably we can split this single functions into couple of smaller ones.

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/BitmapIndexRecordWriter.scala [hunk 4]
```diff
@@ -59,28 +65,65 @@ private[index] class BitmapIndexRecordWriter(
...
+      rbBitMapOut.close()
+    })
+    rbOffsetListBuffer.append(rbOffset + totalRbSize)
+    IndexUtils.writeInt(writer, totalRbSize)
+    sortedKeyList.foreach(sortedKey => {
+      val rb = rowMapRb.get(sortedKey).get
+      rb.runOptimize()
+      val rbWriteBitMapBuf = new ByteArrayOutputStream()
+      val rbBitMapOut = new DataOutputStream(rbWriteBitMapBuf)
+      rb.serialize(rbBitMapOut)
```
Comment (chenghao-intel) [inline src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/BitmapIndexRecordWriter.scala]: I am a little worry about this implementation, ideally, once we defined the layout, the other languages like C/C++ can decode from the index file, however, if we introduced the serialization mechanism, then it's probably a black box for the index file layout. what do you think?

File: src/test/scala/org/apache/spark/sql/execution/datasources/oap/index/BitmapMicroBenchmarkSuite.scala [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+      }
+    }
+    val queryStartTime = System.nanoTime
+    checkAnswer(sql("SELECT * FROM oap_test WHERE a = 15000"),
+      Row(15000, "this is test 15000") :: Nil)
+    val queryEndTime = System.nanoTime
+    // The unit is ms.
+    val queryElapsedTime = (queryEndTime - queryStartTime) / 1000000
+    sql("drop oindex index_bm on oap_test")
+    /* Below result is tested on my local dev machine(Core i7 3.47GHZ with 12 cores, 12GB memory).
```
Comment (chenghao-intel) [inline src/test/scala/org/apache/spark/sql/execution/datasources/oap/index/BitmapMicroBenchmarkSuite.scala]: Cool, I love this benchmark result！

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/BitmapIndexRecordWriter.scala (hunk 4 above)
Comment (lee-lei) [reply src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/BitmapIndexRecordWriter.scala]: The current implementation is just independent. The C/C++ roaring bitmap version is able to decode/deserialize/analyze the bitmap entries one by one. Thanks.

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/BitmapIndexRecordWriter.scala (hunk 3 above)
Comment (lee-lei) [reply src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/BitmapIndexRecordWriter.scala]: I added index layout in the code. I will make sure it's more clear. Thanks.

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/BitMapScanner.scala (hunk 1 above)
Comment (lee-lei) [reply src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/BitMapScanner.scala]: Thanks for pointing out the differences. We are using the roaring bitmap java version. In the roaring bitmap implementation, the iterator is just using short int as the element type for the iterator and cast to unsigned int before returning. Thanks.

**Comment by linhong-intel** [comment]
lgtm. pending on code conflict

**Comment by lee-lei** [comment]
Just fixed. Thanks Linhong!

**Pull request merged**